talk/tools/profiling.tex

```diff
@@ -4,39 +4,228 @@
+   script          Read perf.data and display trace output
+   stat            Run command and gather performance counter statistics
+   top             System profiling tool.
+   version         display the version of perf binary
+   probe           Define new dynamic tracepoints
+   trace           strace inspired tool
+ See 'perf help COMMAND' for more information on a specific command.
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Listing events with \mintinline{bash}{perf list}}
+  { \scriptsize
+    \begin{block}{}
+      \begin{minted}{shell-session}
+$ # List main hardware events
+$ perf list hw
+
+List of pre-defined events (to be used in -e):
+
+  branch-instructions OR branches                    [Hardware event]
+  branch-misses                                      [Hardware event]
+  cache-misses                                       [Hardware event]
+  cache-references                                   [Hardware event]
+  cpu-cycles OR cycles                               [Hardware event]
+  instructions                                       [Hardware event]
+
+$ # List main software/cache events
+$ perf list sw
+$ perf list cache
+
+$ # List all pre-defined metrics
+$ perf list metric
+
+$ # List all currently known events:
+$ perf list
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitle{Counting events with \mintinline{bash}{perf stat}}
+  { \scriptsize
+    \begin{block}{}
+      \begin{minted}{shell-session}
+$ # Standard CPU counter statistics for the specified command:
+$ perf stat <command>
+
+$ # Detailed CPU counter statistics for the specified command:
+$ perf stat -d <command>
+$ perf stat -dd <command>
+
+$ # Top-down microarchitecture analysis for the entire system, for 10s:
+$ perf stat -a --topdown -- sleep 10
+
+$ # L1 cache hit rate reported every 1000 ms for the specified command:
+$ perf stat -e L1-dcache-loads,L1-dcache-load-misses -I 1000 <command>
+
+$ # Instruction per cycle and Instruction-level parallelism, for command:
+$ perf stat -M IPC,ILP -- <command>
+
+$ # Measure GFLOPs system-wide, until Ctrl-C is used to stop:
+$ perf stat -M GFLOPs
+
+$ # Measure cycles and instructions 10 times, report results with stddev:
+$ perf stat -e cycles,instructions -r 10 -- <command>
+      \end{minted}
+    \end{block}
+  }
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitle{Recording profiling information with \mintinline{bash}{perf record}}
+  { \scriptsize
+    \begin{block}{}
+      \begin{minted}{shell-session}
+$ # Sample on-CPU functions for the specified command, at 100 Hertz:
+$ perf record -F 100 -- <command>
+
+$ # Sample CPU stack traces (via frame pointers), at 100 Hertz, for 10s:
+$ perf record -F 100 -g -- sleep 10
+
+$ # Sample stack traces for PID using DWARF to unwind stacks, for 10s:
+$ perf record -p <PID> --call-graph=dwarf -- sleep 10
```

Comment:
Yes, the `sleep` command is only used to give perf the start/stop timings (it's a very common thing to do with perf to use sleep, as there's no other easy way to tell perf to stop otherwise). The profiled process is actually the one given by `<PID>`.

Reply:
And here we suppose that people are at easy with frame-pointers (previous line) and dwarf. That would require another set of slides by itself. Less and less convinced that we should not simplify drastically and give only one slide of examples with one line of each list/stat/record/report

Reply:
I tend to agree with @sponce. Maybe I'm assuming too much prior knowledge that the average student doesn't/won't have. I guess in that case, showing just how to do the simplest case, which is to collect and view a report just using the default of `cycles` for the event is good enough for the course, and we can point people to other sets of slides when more advanced material is needed.

Reply:
But I'm sure HSF people would love to create a full course dedicated to perf. And I promise I would be one of your first students :-)

Reply:
I've given a few talks here and there, so I have many slides on perf (not using LaTeX, though). I could think about converting the material I have into a course on performance analysis, and including other less known tools, like bpftrace, uftrace, bcc, etc.